### DISCUSSION TITLE
ccache cant detect #define's changes with precompiled headers

### DISCUSSION THREAD

**Discussion by shubham4060** [ccache/ccache#1051] [Questions & answers] [answered]
Hi,

I was using PCH with ccache to speed up the compilation. I came across "it can’t detect changes in #defines in the source code because of how preprocessing works in combination with precompiled headers" on ccache manual but since in our project we are using #defines extensively so i was hoping if there is any way ccache would be able to capture the #define changes in a source file?

and as per ccache manual, for the preprocessor mode hash includes "the command line options except options that affect include files (-I, -include, -D, etc; the theory is that these options will change the preprocessor output if they have any effect at all)". But when using ccache with pch I observed in the ccache logs that in preprocessor mode hash consisted of -I(include dirs),-D(preprocessor) options too. can anyone please tell me why is this behavior inconsistent while using PCH?

**Comment by jrosdahl** [comment, marked as answer]
> can anyone please tell me why is this behavior inconsistent while using PCH?

From 708d9110a103bd49437be7bff1e82697fff68d0b: _"The background is that since ccache runs the preprocessor on the header to be precompiled, the preprocessor removes #defines and ccache then hashes the output, which means that changes in #defines are not detected even though the resulting .gch file will have different behavior."_

> i was hoping if there is any way ccache would be able to capture the #define changes in a source file?

The problem is not about detecting changes in #defines when compiling a source file, it's about detecting changes in #defines (NB: inside header files, not command line arguments) when compiling a preprocessed header file. I'm not aware of any way around this.
